Fix part of #6343: Update js files to remove typescript compile errors

ankita240796 opened the pull request:
## Explanation
Fixes part of #6343:  This PR fixes the first three types of errors mentioned in the issue.

ankita240796 committed:
Fix type issues

ankita240796 committed:
fix more errors

ankita240796 committed:
Fix subsequent declaration errors

ankita240796 committed:
Add global third party typings

ankita240796 committed:
Remove assets from migration

ankita240796 committed:
Fix more errors

ankita240796 committed:
Revert changes for test files

seanlip commented:
@ankita240796, it looks like tests are failing. PTAL. Thanks!

ankita240796 commented:
The frontend test were failing because of a generated ID not matching the regex:

```
Expected 'pq6s0df3f' to match /^[a-z0-9]{10}$/.
```

This once happened locally too though I have made no changes to IdGenerationService. After re-running the test, the test passes. I am not sure about why this happens.

@apb7 can you please take a look, Thanks!

seanlip commented:
Oh I have a guess about that. Maybe the ID is being generated numerically and converted into a string? If the initial number is too small then the string will be too short.

seanlip commented on core/templates/dev/head/components/ExplorationCreationService.js:
```diff
@@ -115,7 +115,7 @@ oppia.factory('ExplorationCreationService', [
               parsedResponse.error || 'Error communicating with server.');
             $rootScope.loadingMessage = '';
-            $scope.$apply();
+            $rootScope.$apply();
```
Why?

ankita240796 replied:
$scope is not used in the service and since we are binding values to $rootScope, $apply should be called on rootScope.

seanlip commented on core/templates/dev/head/components/OutcomeDestinationEditorDirective.js:
```diff
@@ -103,10 +103,11 @@ oppia.directive('outcomeDestinationEditor', [
             // has never been rendered at the time this computation is being
             // carried out.
             var stateNames = angular.copy(allStateNames);
+            var stateName;
```
`var stateName = null;` (normally we always initialize)

ankita240796 replied:
Done.

ankita240796 commented:
Should we try something similar as [here](https://stackoverflow.com/questions/1349404/generate-random-string-characters-in-javascript) to generate a random id?

seanlip commented on core/templates/dev/head/domain/collection/EditableCollectionBackendApiServiceSpec.js:
```diff
@@ -94,6 +94,7 @@ describe('Editable collection backend API service', function() {
     function() {
       var successHandler = jasmine.createSpy('success');
       var failHandler = jasmine.createSpy('fail');
+      var collection;
```
= null

ankita240796 replied:
Done.

seanlip commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiServiceSpec.js:
```diff
@@ -132,6 +132,7 @@ describe('Editable exploration backend API service', function() {
     function() {
       var successHandler = jasmine.createSpy('success');
       var failHandler = jasmine.createSpy('fail');
+      var exploration;
```
= null, here and below

ankita240796 replied:
Done.

seanlip commented on core/templates/dev/head/domain/objects/NumberWithUnitsObjectFactory.js:
```diff
@@ -36,7 +36,7 @@ oppia.constant('NUMBER_WITH_UNITS_PARSING_ERRORS', {
     aliases: Other allowed canonical forms of the currency unit.
     front_units: A list of all the currency symbols that are added to the front
-      (like- $, Rs, ₹). Keep it an empty list if no symbol is needed.
+      (like- $, ₹, rupeeSymbol). Keep it an empty list if no symbol is needed.
```
Why change this?

ankita240796 replied:
Missed this while debugging, Fixed.

seanlip commented on core/templates/dev/head/domain/question/EditableQuestionBackendApiServiceSpec.js:
```diff
@@ -121,6 +121,7 @@ describe('Editable question backend API service', function() {
     function() {
       var successHandler = jasmine.createSpy('success');
       var failHandler = jasmine.createSpy('fail');
+      var question;
```
= null

ankita240796 replied:
Done.

seanlip commented on core/templates/dev/head/domain/story/EditableStoryBackendApiServiceSpec.js:
```diff
@@ -124,6 +124,7 @@ describe('Editable story backend API service', function() {
     function() {
       var successHandler = jasmine.createSpy('success');
       var failHandler = jasmine.createSpy('fail');
+      var story;
```
= null

ankita240796 replied:
Done.

seanlip commented on core/templates/dev/head/domain/topic/EditableTopicBackendApiServiceSpec.js:
```diff
@@ -130,6 +130,7 @@ describe('Editable topic backend API service', function() {
     function() {
       var successHandler = jasmine.createSpy('success');
       var failHandler = jasmine.createSpy('fail');
+      var topic;
```
=null

ankita240796 replied:
Done.

seanlip commented on core/templates/dev/head/domain/utilities/UrlInterpolationService.js:
```diff
@@ -119,7 +119,8 @@ oppia.factory('UrlInterpolationService', [
         var escapedInterpolationValues = {};
-        for (var varName in interpolationValues) {
+        var varName;
+        for (varName in interpolationValues) {
```
Instead of declaring a dummy variable varName above can we just use a different name in the while loop below? Maybe something like currentVarName?

ankita240796 replied:
Done.

seanlip commented on core/templates/dev/head/pages/email_dashboard/EmailDashboard.js:
```diff
@@ -17,8 +17,10 @@
 oppia.controller('EmailDashboard', [
-  '$rootDirective', '$scope', 'EmailDashboardDataService', 'UserService',
-  function($rootDirective, $scope, EmailDashboardDataService, UserService) {
+  '$rootDirective', '$rootScope', '$scope', 'EmailDashboardDataService',
```
Actually I don't think $rootDirective is used anywhere.

/cc @oppia/dev-workflow-team FYI

ankita240796 replied:
Yes, it is not used. Shall I remove it then?

seanlip replied:
Yes, please.

/cc @oppia/dev-workflow-team This should be a lint check.

ankita240796 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/ExplorationDataService.js:
```diff
@@ -60,7 +60,10 @@ oppia.factory('ExplorationDataService', [
-      autosaveChangeList: function(changeList, successCallback, errorCallback) {
+      autosaveChangeList: function(
+          changeList,
+          successCallback = function(response) {},
```
Is the unused arg necessary? (I.e. what happens if "response" is removed here?)

ankita240796 replied:
It is necessary since when a call is made to successCallback [here](https://github.com/oppia/oppia/pull/6344/files/46c324c9aed6359b65151d776f27ba140d38e1d5#diff-f5cf086a2061475d6e1fc4cbdc574a4cL76), we pass in response as a parameter.

seanlip replied:
So, just to confirm, removing it results in errors?

ankita240796 replied:
Yes, it does (signature not matches error).

seanlip commented on core/templates/dev/head/pages/exploration_editor/ExplorationDataService.js:
```diff
@@ -72,13 +75,9 @@ oppia.factory('ExplorationDataService', [
-            successCallback(response);
-          }
+          successCallback(response);
         }, function() {
```
You could probably just replace these three lines with `errorCallback`.

ankita240796 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/translation_tab/StateTranslationDirective.js:
```diff
@@ -80,7 +80,7 @@ oppia.directive('stateTranslation', [
               return;
             }
-            var activecontentId = null;
+            var activeContentId;
```
= null

ankita240796 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/ConversationSkinDirective.js:
```diff
@@ -615,6 +617,10 @@ oppia.directive('conversationSkin', [
 
           $rootScope.$on('playerStateChange', function(evt, newStateName) {
+            var isLoggedIn = false;
+            UserService.getUserInfoAsync().then(function(userInfo) {
```
This might be problematic, because it happens asynchronously. What if the async call takes a long time and the stuff below executes before this function call returns?

ankita240796 replied:
Used `$scope.isLoggedIn` instead, I missed it earlier.

seanlip commented on core/templates/dev/head/pages/exploration_player/ExtractImageFilenamesFromStateService.js:
```diff
@@ -124,6 +124,7 @@ oppia.factory('ExtractImageFilenamesFromStateService', [
 
       var imageTagList = dummyElement.getElementsByTagName(
         'oppia-noninteractive-image');
+      var i;
```
Put i in the for loop?

ankita240796 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/PretestEngineService.js:
```diff
@@ -196,6 +196,7 @@ oppia.factory('PretestEngineService', [
           return;
         }
 
+        var newState;
```
= null

ankita240796 replied:
Done.

seanlip commented on core/templates/dev/head/pages/library/Library.js:
```diff
@@ -156,6 +156,7 @@ oppia.controller('Library', [
         // elements flush left.
         // Transforms the group names into translation ids
         $scope.leftmostCardIndices = [];
+        var i;
```
Put var in the for loop instead? Ditto below. You could use different vars too (i for one, j for another).

ankita240796 replied:
Done.

seanlip commented on core/templates/dev/head/pages/story_editor/main_editor/StoryNodeEditorDirective.js:
```diff
@@ -43,6 +43,7 @@ oppia.directive('storyNodeEditor', [
           var _recalculateAvailableNodes = function() {
             $scope.newNodeId = null;
             $scope.availableNodes = [];
+            var i;
```
Ditto

ankita240796 replied:
Done.

seanlip commented on core/tests/protractor_utils/ExplorationEditorMainTab.js:
```diff
@@ -431,6 +431,7 @@ var ExplorationEditorMainTab = function() {
       destName, createNewDest, refresherExplorationId) {
     expect(destName === null && createNewDest).toBe(false);
 
+    var targetOption;
```
= null

ankita240796 replied:
Done.

seanlip commented on core/tests/protractor_utils/forms.js:
```diff
@@ -85,6 +85,7 @@ var GraphEditor = function(graphInputContainer) {
         by.css('.protractor-test-Delete-button'));
       deleteButton.click();
       // Sample graph comes with 3 vertices.
+      var i;
```
ditto

ankita240796 replied:
Done.

seanlip commented on extensions/interactions/CodeRepl/CodeReplPredictionServiceSpec.js:
```diff
@@ -132,6 +132,7 @@ describe('CodeRepl prediction service', function() {
 
       // To keep things simple, we will calculate accuracy score
       // and not F1 score.
+      var predictedAnswerGroup;
```
= null

ankita240796 replied:
Done.

seanlip commented on extensions/interactions/LogicProof/static/js/shared.js:
```diff
@@ -559,10 +559,9 @@ var logicProofShared = (function() {
    *         to what the user intended.
    */
   var assignTypesToExpression = function(untypedExpression, possibleTopTypes,
```
Generally, if contents of parens span more than a line, break after '('. Ditto below.

ankita240796 replied:
Done.

seanlip commented on extensions/interactions/TextInput/TextInputPredictionServiceSpec.js:
```diff
@@ -34,6 +34,7 @@ describe('Text Input Prediction Service', function() {
       var classifierData = getJSONFixture('text_input_classifier_data.json');
       var trainingData = getJSONFixture('text_classifier_results.json');
       var correctPredictions = 0, totalAnswers = 0;
+      var predictedAnswerGroup;
```
= null

ankita240796 replied:
Done.

seanlip commented on extensions/interactions/TextInput/TextInputPredictionServiceSpec.js:
```diff
@@ -63,6 +64,7 @@ describe('Text Input Prediction Service', function() {
 
       // To keep things simple, we will calculate accuracy score
       // and not F1 score.
+      var predictedAnswerGroup;
```
= null

ankita240796 replied:
Done.

seanlip commented on globals.d.ts:
```diff
@@ -0,0 +1,5 @@
+declare var MIDI: any;
```
Make sure to file issue for resolving this (please link to it in your reply to this comment)

ankita240796 replied:
Done. Filed #6351

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -1344,7 +1343,16 @@ def _validate_and_parse_js_files(all_files):
         content = FileCache.read(filename)
-        parsed_js_files[filename] = parser.parse(content)
+
+        content_with_no_non_ascii_characters = ''
```
Hm, this seems a bit dodgy -- what's going on here, and why is this necessary?

ankita240796 replied:
Parsing js files with esprima raises an illegal token exception when non ascii characters like `₹` are present in the file. So, before parsing file contents, all non ascii characters are removed.

seanlip replied:
No, don't do that. Find a way to parse the complete file without removing random characters from it -- maybe you need to encode it properly, or change a setting in esprima, or something. You might consider asking the esprima folks for help if needed, but in general you shouldn't be changing the "meaning" of the file when linting.

ankita240796 replied:
Sent you a mail outlining instructions to reproduce the error.

ankita240796 replied:
I was trying to convert characters to their respective unicode form as we discussed but it was giving unexpected end of data error. While searching through, I found a better fix [here](https://stackoverflow.com/a/10487617). The read operation was just returning bytes and decoding the bytes fixed the issue!

seanlip reviewed (changes requested):
Thanks, @ankita240796! Taken a full pass, PTAL.

Also, other reviewers -- please don't worry about reviewing this. I've done a pretty thorough pass and can take this entire review.

seanlip commented:
> Should we try something similar as [here](https://stackoverflow.com/questions/1349404/generate-random-string-characters-in-javascript) to generate a random id?

That could work. Could you point to the code that is causing this? If it's in tests, that should be fine. If it's in the main code we'll need to check it a bit more carefully.

ankita240796 committed:
Address review comments

ankita240796 commented:
Hi @seanlip, I have addressed the review comments, PTAL.

Regarding your query here,

> That could work. Could you point to the code that is causing this? If it's in tests, that should be fine. If it's in the main code we'll need to check it a bit more carefully.

The code where error is raised is in the test [here](https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/services/IdGenerationServiceSpec.js#L28) and the function used for id generation is defined [here](https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/services/IdGenerationService.js#L21).

Thanks!

ankita240796 committed:
Add type definition for visualization

ankita240796 committed:
Fix test

ankita240796 committed:
fix left out error

seanlip commented on core/templates/dev/head/pages/exploration_editor/ExplorationDataService.js:
```diff
@@ -72,14 +75,9 @@ oppia.factory('ExplorationDataService', [
-        });
+          successCallback(response);
+        },
+        errorCallback());
```
errorCallback, not errorCallback(). The argument is a function, not the result of the function's execution.

ankita240796 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/PlayerPositionServiceSpec.js:
```diff
@@ -19,7 +19,7 @@
   beforeEach(module('oppia'));
 
-  var pts, pps, scof;
+  var pts = null, pps = null, scof = null;
```
FYI we typically always declare on separate lines.

ankita240796 replied:
Done.

ankita240796 committed:
Merge branch 'develop' into fix-type

seanlip reviewed (changes requested):
Looking good, just a few more comments.

For the generateNewId() issue I think the diagnosis I gave earlier is correct. One simple fix would be to just regenerate in a while loop until the resulting ID is 10 chars long.

ankita240796 committed:
Resolve conflicts

ankita240796 committed:
Address review comments

ankita240796 committed:
Merge branch 'fix-type' of github.com:ankita240796/oppia into fix-type

ankita240796 commented:
Hi @seanlip, I have addressed the review comments, PTAL. Thanks!

ankita240796 committed:
Fix left out issue

seanlip commented on core/templates/dev/head/pages/exploration_editor/ExplorationDataService.js:
```diff
@@ -87,7 +87,7 @@ oppia.factory('ExplorationDataService', [
         }, function() {
           if (errorCallback) {
-            errorCallback();
+            errorCallback;
```
No -- at this point you need to execute the function. So keep the `()`.

ankita240796 replied:
Sorry, my bad. Fixed.

seanlip reviewed (commented):
Only one note, otherwise LGTM once the inline comment and the esprima issue get resolved. Thanks!

ankita240796 committed:
Fix braces

ankita240796 committed:
Fix unicode errors

ankita240796 committed:
Merge branch 'fix-type' of github.com:ankita240796/oppia into fix-type

ankita240796 commented:
Hi @seanlip, esprima issue is fixed now! PTAL, Thanks!

seanlip reviewed (approved):
Looks great, @ankita240796. Thanks!
